### PR TITLE
Specify a custom logo file to use instead of the Nintendo logo

### LINKED ISSUE #1398: [Feature request] Support arbitrary logos in the fixed header
RGBFIX can fill in the Nintendo logo in the header with its usual 48 bytes. However, some alternative consoles use a different logo, such as the Analogue Pocket. Without hard-coding these third-parties' logos (which could be taken as endorsing them), we can still let the user supply an arbitrary logo from a file.

Logo files could just be 48-byte binary blobs. But, we could also allow 48x8-px (6x1-tile) 1bpp input, and convert that to the weird header logo encoding.

Nintendo:
```
██░░░██░██░░░░░░░░░░░░░░░░░░░░░░░░░░░░░██░░░░░░░
███░░██░██░░░░░░░░██░░░░░░░░░░░░░░░░░░░██░░░░░░░
███░░██░░░░░░░░░░████░░░░░░░░░░░░░░░░░░██░░░░░░░
██░█░██░██░██░██░░██░░████░░██░██░░░█████░░████░
██░█░██░██░███░██░██░██░░██░███░██░██░░██░██░░██
██░░███░██░██░░██░██░██████░██░░██░██░░██░██░░██
██░░███░██░██░░██░██░██░░░░░██░░██░██░░██░██░░██
██░░░██░██░██░░██░██░░█████░██░░██░░█████░░████░
```

Analogue:
```
░░░░██░░░░░░░█░░░░░░░░░░░░░█░░░░░░░░░░░░░░░░░░░░
░░░████░░░░░░█░░░░░░░░░░░░░█░░░░░░░░░░░░░░░░░░░░
░░░████░░░░░█░█░░███░░░██░░█░░██░░░███░█░░█░░██░
░░░░████░░░░█░█░░█░░█░░░░█░█░█░░█░█░░█░█░░█░█░░█
░██░████░░░░███░░█░░█░░███░█░█░░█░█░░█░█░░█░████
████░████░░█░░░█░█░░█░█░░█░█░█░░█░░███░█░░█░█░░░
████░████░░█░░░█░█░░█░░███░█░░██░░█░░█░░███░░██░
░██░░░██░░░░░░░░░░░░░░░░░░░░░░░░░░░██░░░░░░░░░░░
```

### PR DESCRIPTION
**Pull request by Rangi42**
Fixes #1398

The new test case is the Analogue Pocket logo.

`xxd custom-logo.1bpp`:
```
00000000: 0c1e 1e0f 6ff7 f763 0404 0a0a 0e91 9100  ....o..c........
00000010: 0000 7148 494a 4900 1010 9354 d454 d300  ..qHIJI....T.T..
00000020: 0000 1da5 a59d 2418 0000 2629 2f28 e600  ......$...&)/(..
```

`xxd custom-logo.gb`:
```
00000100:           0110 ceef 0000 44aa 0074 0018      ......D..t..
00000110: 1195 0034 001a 00d5 0022 0069 6ff6 f773  ...4.....".io..s
00000120: 0990 e110 4440 9a90 d5d0 4430 a921 5d48  ....D@....D0.!]H
00000130: 22e0 f860                                "..`.
```

### CODE REVIEW

File: src/fix/main.cpp [hunk 1]
```diff
@@ -1451,6 +1445,55 @@ int main(int argc, char *argv[]) {
 	argv += musl_optind;
 	bool failed = nbErrors;
 
+	if (logoFilename && (fixSpec & (FIX_LOGO | TRASH_LOGO))) {
```
Comment (Rangi42) [inline src/fix/main.cpp:1448]: @ISSOtm Should we read the custom logo file unconditionally, or only if we'll be using it? If you pass `-L nonexistent.1bpp` without passing `-f l` (or `-f L`, or `-v`), should rgbfix give an error?

Comment (ISSOtm) [reply src/fix/main.cpp:1448]: I think it's fair to go either way. So I'd break that tie by whatever is the simplest for us. (Thus, unconditionally reading it.)

File: test/fix/test.sh [hunk 2]
```diff
@@ -39,6 +39,7 @@ tryCmp () {
 
 runTest () {
 	flags="$(head -n 1 "$2/$1.flags")" # Allow other lines to serve as comments
+	flags=$(echo "$flags" | sed "s#-L #-L $src/#g") # Prepend src directory to logo file
```
Comment (ISSOtm) [inline test/fix/test.sh:42]:
```suggestion
	flags=$(
		head -n 1 "$2/$1.flags" | # Allow other lines to serve as comments
		sed "s#-L #-L ${src//#/\\#}/#g" # Prepend src directory to logo file
	)
```
Using only a single expansion (which also avoids `echo` interpreting any of the flags as its own&mdash;`printf` avoids that), and also escaping any hashes that might have snuck into `$src`.

**Review by ISSOtm — CHANGES_REQUESTED** [review]
Great!

**Comment by Rangi42** [comment]
Copying comments here that I put on the wrong PR :sweat_smile: :

> > It occurs to me that another way to do this would be to make `-L custom.1bpp` imply `-f l` (same as how `-v` implies `-f lhg`), and not allow "trashing" with a custom logo since the particular trash values don't matter (in other words, `-f L` passed after `-L custom.1bpp` would warn about overriding, and then the logo would be fixed to the inverse of the Nintendo one, ignoring the overridden custom logo). @ISSOtm Would that be preferable?
> 
> That would break if someone's custom logo is, for whatever reason, the "trashed" Nintendo logo. (It seems unlikely, but is still a possibility.)
> 
> `-v` does not quite *imply* `-f lhg` as much as is a *shorthand* for it. Arguably also legacy, but `lhg` is also far more common than any other combination.
> 
> I think one could also argue that `-f l` should also imply `-f h` (because the header checksum also needs to be recalculated), and likewise `-f h` implying `-f g`.
But we're not doing that, and I think the cost of requiring people to explicitly specify all three is minimal enough that it is a sane approach to keep. (Plus it has less implicit behaviour, and I agree with at least *that* verse of the Zen of Python.)
> 
> Lastly, I can see someone doing `make RGBFIX="rgbfix -L logo.1bpp`" and that calling `rgbfix -L logo.1bpp -f gh rom.gb` at some point.
> 
> Given all of the above, I think that `-L` should not imply `-f l`. And preferably, `-f L` should take `-L` into account. If reasonably possible.

File: src/fix/main.cpp [hunk 3]
```diff
@@ -1451,6 +1445,55 @@ int main(int argc, char *argv[]) {
...
+			logo[logoWidth + i + 2] = lows(i + 2);
+			logo[logoWidth + i + 3] = lows(i + 3);
+		}
+	} else {
+		memcpy(logo, nintendoLogo, sizeof(nintendoLogo));
+	}
+	if (fixSpec & TRASH_LOGO) {
+		for (uint16_t i = 0; i < sizeof(logo); i++)
+			logo[i] = 0xFF ^ logo[i];
+	}
```
[the remark this answers is not in the document]
Comment (Rangi42) [reply src/fix/main.cpp:1495]: I considered it, but noticed that the rest of rgbfix only breaks out separate functions when they're reused. A lot of behavior is consolidated in just the `main`, `processFilename`, and `processFile` functions.

File: man/rgbfix.1 [hunk 4]
```diff
@@ -104,6 +105,9 @@ Set the new licensee string
 .Pq Ad 0x144 Ns \(en Ns Ad 0x145
 to a given string.
 If it's longer than 2 chars, it will be truncated, and a warning emitted.
+.It Fl L Ar logo_file , Fl \-logo Ar logo_file
+Specify a logo file to use instead of the official Nintendo logo.
+The file must be a 48-byte 1bpp image (48 pixels wide and 8 pixels tall).
```
Comment (ISSOtm) [inline man/rgbfix.1:110]: Calling it an "image" will probably lead people to try feeding that option PNG files.

```suggestion
The file must be 48 bytes of 1bpp tile data (the source image should be 48 pixels wide and 8 pixels tall).
```

File: src/asm/main.cpp
Comment (ISSOtm) [inline src/asm/main.cpp:1]: Not too sure what this change achieves?

**Review by ISSOtm — CHANGES_REQUESTED** [review]

**Review by ISSOtm — APPROVED** [review]

**Pull request merged**